app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -550,22 +556,108 @@ protected void deliverError(CommCareSetupActivity receiver,
+        }
+        if (profileLink != null) {
+            // we found a valid profile link, either start install automatically
+            // or move to READY_TO_INSTALL state
+            Log.v("install", "Performing SMS install with link : " + incomingRef);
+            incomingRef = profileLink;
+            if (installAutomatically) {
+                startResourceInstall();
+            } else {
+                uiState = UiState.READY_TO_INSTALL;
+                Toast.makeText(this, Localization.get("menu.sms.ready"), Toast.LENGTH_LONG).show();
+            }
+        } else if(installAutomatically) {
+            // only notify if this was manually triggered, since most people won't use this
+            Toast.makeText(this, Localization.get("menu.sms.not.found"), Toast.LENGTH_LONG).show();
+        }
+    }
+
+
+
+    /**
+     * Scan the SMS inbox, looking for messages that meet the expected install format,
+     * and if found and verified return the discovered install link. Current behavior will search
+     * backwards from the most recent text, returning the first discovered valid link
+     * @return the verified install link, null if none found
+     * @throws SignatureException if we discovered a valid-looking message but could not verifyMessageSignatureHelper it
+     */
+    private String scanSMSLinks() throws SignatureException{
+        // http://stackoverflow.com/questions/11301046/search-sms-inbox
+        final Uri SMS_INBOX = Uri.parse("content://sms/inbox");
+        Cursor cursor = getContentResolver().query(SMS_INBOX, null, null, null, "date desc");
+        try {
+            while (!cursor.isAfterLast()) {
+                String textMessageBody = cursor.getString(cursor.getColumnIndex("body"));
+                if (textMessageBody.contains("ccapp:") && textMessageBody.contains("signature:")) {
+                    String installLink = parseAndVerifySMS(textMessageBody);
+                    if(installLink != null){
+                        return installLink;
+                    }
+                }
+            }
+        }
+        finally{
+            cursor.close();
+        }
+        return null;
+    }
+
+    /**
+     *
+     * @param text the parsed out text message in the expected link/signature format
+     * @return the download link if the message was valid and verified, null otherwise
+     * @throws SignatureException if we discovered a valid-looking message but could not verifyMessageSignatureHelper it
+     */
+    private String parseAndVerifySMS(String text) throws SignatureException {
+        // parse out the app link and signature. We assume there is a space after ccapp: and
+        // signature: and that the end of the signature is the end of the text content
+        String downloadLink = text.substring(text.indexOf("ccapp:") + 7, text.indexOf(","));
+        String signature = text.substring(text.indexOf("signature:") + 11);

```

Comment:
This does match the format I shared in the example, but now that I think about it, there's a couple weaknesses in that 1) it exposes another link to the user which might confuse them and make them click on the wrong link, and 2) the user might think it's junk and delete it which would cause problems too.  What do you think about changing the format to be:

`[COMMCARE SMS - DO NOT DELETE] <base 64 encoded message>`, and once you decode `<base 64 encoded message>`, you get: `ccapp: <applink> signature: <signature>`, with `<applink>` being the same app link and now `<signature>` is the signature in binary format.  Also note this gets rid of the comma which I think makes things cleaner so that there's just one space as padding.